### PR TITLE
[im] Add command obj to all cluster callbacks

### LINKED ISSUE #5098: [IM] Add Command object and endpoint id to params of current command handlers
**Problem**

Then it will be better to fix the current commands handlers to have a new parameter for it in a separate PR.
Please remove `src/app/clusters/im-compatibility`, it introduce a new generated folder in `src/` and creates a level of indirection that does not seems really needed.

_Originally posted by @vivien-apple in https://github.com/project-chip/connectedhomeip/pull/5032#discussion_r585391808_

Currently, the commands are using a global emberApsFrame & emberAfCluster object, this does not meet the needs of the requirement of CHIP datamodel.

The IME provideds a Command* object, and the command handlers in the future should be able to make use of them.

### PR DESCRIPTION
**Pull request by erjiaqing**
#### Problem

Currently, the commands are using a global emberApsFrame & emberAfCluster object, this does not meet the needs of the requirement of CHIP datamodel.

The IME provideds a Command* object, and the command handlers in the future should be able to make use of them.

 #### Summary of Changes

- Replace  `bool emberAf(.+)Cluster(.+)Callback(` to `bool emberAf$1Cluster$2Callback(chip::app::Command * apCommandObj`

- Update codegen

- Fix build for some mismatch

- Run `./scripts/tools/zap_regen_all.py && for x in $(gd --name-only | grep -ve '(.zap$)|(.py$)|(.gn$)|(.zapt$)'); do clang-format -i $x; done && ./gn_build.sh chip_build_tests=false && git add src/ examples/ && gc -m "Run Codegen" && gp -f origin im-command-obj`

 Fixes #5098

### CODE REVIEW

File: examples/all-clusters-app/esp32/main/DeviceCallbacks.cpp [hunk 1]
```diff
@@ -174,7 +175,7 @@ void DeviceCallbacks::OnIdentifyPostAttributeChangeCallback(EndpointId endpointI
     return;
 }
 
-bool emberAfBasicClusterMfgSpecificPingCallback(void)
+bool emberAfBasicClusterMfgSpecificPingCallback(chip::app::Command *)
```
Comment (vivien-apple) [inline examples/all-clusters-app/esp32/main/DeviceCallbacks.cpp:178]: Can we add a default name, for this one and for all the other occurrences where there is no default name.

File: src/app/clusters/account-login-server/account-login-server.cpp [hunk 2]
```diff
@@ -38,16 +38,17 @@
  *******************************************************************************
  ******************************************************************************/
 
+#include <app/Command.h>
 #include <app/util/af.h>
 
-bool emberAfAccountLoginClusterGetSetupPINCallback(unsigned char *)
+bool emberAfAccountLoginClusterGetSetupPINCallback(chip::app::Command * apCommandObj, unsigned char *)
```
Comment (vivien-apple) [inline src/app/clusters/account-login-server/account-login-server.cpp:44]: `ap` prefix is not the style that is used in those files. Please call it `commandObj` or something like that to stay consistent with the style used in those files. For this occurrence and for the others too.

File: src/app/clusters/basic/basic.cpp [hunk 3]
```diff
@@ -20,6 +20,7 @@
 
 #include <platform/CHIPDeviceLayer.h>
 
+#include <app/Command.h>
```
Comment (vivien-apple) [inline src/app/clusters/basic/basic.cpp:23]: Is it needed ?

File: src/app/clusters/descriptor/descriptor.cpp [hunk 4]
```diff
@@ -21,6 +21,7 @@
  ***************************************************************************/
 
 #include "af.h"
+#include <app/Command.h>
```
Comment (vivien-apple) [inline src/app/clusters/descriptor/descriptor.cpp:24]: Is it needed ?

File: src/app/zap-templates/partials/im_command_handler_cluster_commands.zapt [hunk 5]
```diff
@@ -79,7 +79,7 @@ if (CHIP_NO_ERROR == TLVError && CHIP_NO_ERROR == TLVUnpackError && {{zcl_comman
 {
 {{/if}}
 // TODO(#5098) We should pass the Command Object and EndpointId to the cluster callbacks.
-emberAf{{asCamelCased parent.name false}}Cluster{{asCamelCased name false}}Callback({{#zcl_command_arguments}}{{#if (isCharString type)}}const_cast<uint8_t*>({{asSymbol label}}){{else}}{{asSymbol label}}{{/if}}{{#unless (isLastElement index count)}}, {{/unless}}{{/zcl_command_arguments}});
+emberAf{{asCamelCased parent.name false}}Cluster{{asCamelCased name false}}Callback(apCommandObj{{#zcl_command_arguments}}, {{#if (isCharString type)}}const_cast<uint8_t*>({{asSymbol label}}){{else}}{{asSymbol label}}{{/if}}{{/zcl_command_arguments}});
```
Comment (vivien-apple) [inline src/app/zap-templates/partials/im_command_handler_cluster_commands.zapt:82]: Same comment as for the argument name in the `src/app/clusters` files, `ap` prefix is not the style of the the code there.

File: src/app/zap-templates/templates/app/CHIPClientCallbacks-src.zapt [hunk 6]
```diff
@@ -700,7 +701,7 @@ bool emberAfDiscoverCommandsReceivedResponseCallback(ClusterId clusterId, uint16
 {{#if (isStrEqual clusterName parent.name)}}
 {{#if (isCommandAvailable parent.side incoming outgoing commandSource name)}}
 {{#if (isStrEndsWith name "Response")}}
-bool emberAf{{asCamelCased parent.name false}}Cluster{{asCamelCased name false}}Callback({{#zcl_command_arguments}}{{asUnderlyingZclType type}} {{asSymbol label}}{{#unless (isLastElement index count)}}, {{/unless}}{{/zcl_command_arguments}})
+bool emberAf{{asCamelCased parent.name false}}Cluster{{asCamelCased name false}}Callback(chip::app::Command * apCommandObj{{#zcl_command_arguments}}, {{asUnderlyingZclType type}} {{asSymbol label}}{{/zcl_command_arguments}})
```
Comment (vivien-apple) [inline src/app/zap-templates/templates/app/CHIPClientCallbacks-src.zapt:704]: Same as above.

File: src/app/zap-templates/templates/app/callback.zapt [hunk 7]
```diff
@@ -119,7 +120,7 @@ void emberAf{{asCamelCased name false}}Cluster{{asCamelCased side false}}TickCal
 {{/if}}
 */
 
-bool emberAf{{asCamelCased parent.name false}}Cluster{{asCamelCased name false}}Callback({{#zcl_command_arguments}}{{#if (isOctetString type)}}chip::ByteSpan{{else}}{{asUnderlyingZclType type}}{{/if}} {{asSymbol label}}{{#unless (isLastElement index count)}}, {{/unless}}{{/zcl_command_arguments}});
+bool emberAf{{asCamelCased parent.name false}}Cluster{{asCamelCased name false}}Callback(chip::app::Command * apCommandObj{{#zcl_command_arguments}}, {{#if (isOctetString type)}}chip::ByteSpan{{else}}{{asUnderlyingZclType type}}{{/if}} {{asSymbol label}}{{/zcl_command_arguments}});
```
Comment (vivien-apple) [inline src/app/zap-templates/templates/app/callback.zapt:123]: Same as above.

**Comment by erjiaqing** [comment]
@vivien-apple Thanks, all comment resolved.

**Review by vivien-apple — APPROVED** [review]
Looks good to me. The strange thing is that this PR seems to contains #6119 too.

**Comment by erjiaqing** [comment]
> Looks good to me. The strange thing is that this PR seems to contains #6119 too.

Yes, I will rebase after #6119

**Review by msandstedt — APPROVED** [review]

**Review by andy31415 — APPROVED** [review]

**Review by woody-apple — APPROVED** [review]

**Pull request merged**